elastic/eui
# Add caching mechanism to EuiIcon

## Summary

While working on Lens, we encountered fairly significant performance problems due to rendering churn when loading large lists of icons. The churn was caused by the fact that each icon renders at least twice (the initial "empty" mode, and then when the promise resolves, the result of the resolve). 

This PR adds a cache so that the initial render can often be the final render. It caches the result of the promise, and no longer does a re-render for cached icons.

We can probably add a further optimization, which is to use React's memoization feature to avoid diffs for icons whose props haven't changed.

Regarding this implementation, I don't love the global cache, or the way I had to expose things to the test infrastructure, though. Feedback is much appreciated.

## Manual testing

I manually simulated the Lens situation by modifying `src-docs/src/views/icon/icon_example.js` (see the JS snippet below). It performs a "lazy" scrolling mechanism similar to the one we're using in Lens. As you scroll using the original icon implementation, you'll notice the scroll jitters a lot, but with caching, the jittering is much less noticeable.

```js
import React, { useState, useEffect } from 'react';

import { renderToHtml } from '../../services';

import {
  EuiCode,
  EuiIcon,
  EuiText,
  EuiSpacer,
} from '../../../../src/components';
import { palettes } from '../../../../src/services';

const iconHtmlWarning = () => (
  <p>
    HTML preview disabled. Icons use SVG and are not usable usable without React
    unless you load the icons manually through a separate asset loader.
  </p>
);

const PAGINATION_SIZE = 50;
const iconNames = ['string', 'number', 'tokenBoolean', 'calendar'];
const allFields = (() => {
  const result = [];
  for (let i = 0; i < PAGINATION_SIZE * 100; ++i) {
    result.push({
      icon: iconNames[i % iconNames.length],
      text: `icon ${iconNames[i % iconNames.length]} ${i}`,
    });
  }
  result.push({
    icon: 'exit',
    text: 'THE END! FIN!',
  });
  return result;
})();
function stringToNum(s) {
  // tslint:disable-next-line:no-bitwise
  return Array.from(s).reduce((acc, ch) => acc + ch.charCodeAt(0), 1);
}

export function LongScroll(props) {
  const [pageSize, setPageSize] = useState(PAGINATION_SIZE);
  const filteredFields = allFields.slice(0, pageSize);
  const [scrollContainer, setScrollContainer] = useState();

  const lazyScroll = () => {
    if (scrollContainer) {
      const nearBottom =
        scrollContainer.scrollTop + scrollContainer.clientHeight > scrollContainer.scrollHeight * 0.9;
      console.log(
        scrollContainer.scrollTop, scrollContainer.clientHeight, scrollContainer.scrollHeight,

        scrollContainer.scrollTop + scrollContainer.clientHeight, ' >= ', scrollContainer.scrollHeight * 0.9);
      if (nearBottom) {
        setPageSize(Math.min(pageSize * 1.5, allFields.length));
      }
    }
  };

  useEffect(
    () => {
      if (scrollContainer) {
        scrollContainer.scrollTop = 0;
        setPageSize(PAGINATION_SIZE);
        lazyScroll();
      }
    },
    []
  );

  return (
    <div className="lnsFieldListPanel__list">
      <div
        className="lnsFieldListPanel__overflow"
        style={{
          height: '450px',
          overflow: 'auto',
        }}
        ref={el => {
          if (el && !el.dataset.dynamicScroll) {
            el.dataset.dynamicScroll = 'true';
            setScrollContainer(el);
          }
        }}
        onScroll={lazyScroll}
      >
        {filteredFields
          .map(field => {
            const { colors } = palettes.euiPaletteColorBlind;
            const colorIndex = stringToNum(field.icon) % colors.length;
            return (
              <div key={field.text}>
                <EuiIcon type={field.icon} color={colors[colorIndex]} />;
                {field.text}
              </div>
            );
          })}
      </div>
    </div>
  );
}


export const IconExample = {
  title: 'Icons',
  intro: (
    <div>
      <EuiText>
        <p>
          <EuiCode>EuiIcon</EuiCode> is a handy component for using our custom
          glyphs and logos. The <EuiCode>type</EuiCode> prop accepts either an
          enumerated name from one of the sets below, a location to a custom SVG
          asset, or a React Element.
        </p>
        <p>
          When using custom SVGs please{' '}
          <strong>remove all fill attributes</strong> on the SVG and utilize the
          CSS helpers if you have complex logos that need to work with theming.
        </p>
      </EuiText>
      <EuiSpacer />
      <LongScroll />
    </div>
  ),
  sections: [],
};
```
